crates/noirc_evaluator/src/ssa/ssa_gen/context.rs

```diff
@@ -537,7 +537,14 @@ impl<'a> FunctionContext<'a> {
     /// version, as it is only needed for recursion.
     pub(super) fn extract_current_value(&mut self, lvalue: &ast::LValue) -> LValue {
         match lvalue {
-            ast::LValue::Ident(ident) => LValue::Ident(self.ident_lvalue(ident)),
+            ast::LValue::Ident(ident) => {
+                let (reference, should_auto_deref) = self.ident_lvalue(ident);
+                if should_auto_deref {
+                    LValue::Dereference { reference }
+                } else {
+                    LValue::Ident
+                }
+            }
```

Comment:
Instead of having both identifiers and dereferences being possibly assigned to, I thought it was simple to always only assign to LValue::Dereferences. Note that in the case of e.g. `let x = &mut 0`, the identifier `x` itself is not mutable, so the reverse of always only assigning to identifiers is not possible.